crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,359 @@ impl<'a> FileNoqaDirectives<'a> {
+    /// Tracks whether we are lexing in a context with a missing delimiter
+    /// e.g. at `C` in `F401C402`.
+    missing_delimiter: bool,
+    /// Rule codes collected during lexing
+    codes: Vec<Code<'a>>,
+}
+
+impl<'a> NoqaLexer<'a> {
+    fn new(source: &'a str, range: TextRange) -> Self {
+        Self {
+            line: &source[range],
+            offset: range.start(),
+            cursor: Cursor::new(&source[range]),
+            warnings: Vec::new(),
+            missing_delimiter: false,
+            codes: Vec::new(),
+        }
     }
 
-    /// Lex a specific character, or return `None` if the character is not the first character in
-    /// the line.
-    #[inline]
-    fn lex_char(line: &str, c: char) -> Option<&str> {
-        let mut chars = line.chars();
-        if chars.next() == Some(c) {
-            Some(chars.as_str())
-        } else {
-            None
+    /// Initialize [`NoqaLexer`] at offset.
+    ///
+    /// The offset should be at or after the `noqa` literal.
+    fn starts_at(offset: TextSize, source: &'a str) -> Self {
+        let range = TextRange::new(offset, source.text_len());
+        Self::new(source, range)
+    }
+
+    /// Collect codes in `noqa` comment.
+    ///
+    /// The `comment_start` is the offset in the source of the
+    /// beginning of the `noqa` comment, which is earlier
+    /// than the offset that marks the beginning of the codes.
+    fn lex(mut self, comment_start: TextSize) -> Result<NoqaLexerOutput<'a>, LexicalError> {
+        if self.cursor.first().is_ascii_whitespace()
+            || self.cursor.first() == '#'
+            || self.cursor.is_eof()
+        {
+            return Ok(NoqaLexerOutput {
+                warnings: self.warnings,
+                directive: Directive::All(All {
+                    range: TextRange::new(comment_start, self.offset),
+                }),
+            });
         }
+
+        if self.cursor.first() != ':' {
+            return Err(LexicalError::InvalidSuffix);
+        }
+
+        self.cursor.bump();
+        self.lex_codes()?;
+
+        let Some(last) = self.codes.last() else {
+            return Err(LexicalError::MissingCodes);
+        };
+
+        Ok(NoqaLexerOutput {
+            warnings: self.warnings,
+            directive: Directive::Codes(Codes {
+                range: TextRange::new(comment_start, last.range.end()),
+                codes: self.codes,
+            }),
+        })
     }
 
-    /// Lex the "flake8" prefix of a `noqa` directive.
-    #[inline]
-    fn lex_flake8(line: &str) -> Option<&str> {
-        line.strip_prefix("flake8")
+    fn lex_codes(&mut self) -> Result<(), LexicalError> {
+        // SAFETY: Every call to `lex_code` advances the cursor at least once.
+        while !self.cursor.is_eof() {
+            self.lex_code()?;
+        }
+        Ok(())
+    }
+
+    fn lex_code(&mut self) -> Result<(), LexicalError> {
+        self.cursor.start_token();
+        match self.cursor.first() {
```

Comment:
Nit: I prefer to always use `bump` because it ensures that:

a) The lexer always progresses
b) It handles the EOF case